src/Plugin/KeyType/ApigeeAuthKeyType.php

```diff
@@ -0,0 +1,135 @@
+ *     "plugin" = "apigee_auth_input"
+ *   },
+ *   multivalue = {
+ *     "enabled" = true,
+ *     "fields" = {
+ *       "auth_type" = {
+ *         "label" = @Translation("Authentication type"),
+ *         "required" = true
+ *       },
+ *       "organization" = {
+ *         "label" = @Translation("Organization"),
+ *         "required" = true
+ *       },
+ *       "username" = {
+ *         "label" = @Translation("Username"),
+ *         "required" = true
+ *       },
+ *       "password" = {
+ *         "label" = @Translation("Password"),
+ *         "required" = true
+ *       },
+ *       "endpoint" = {
+ *         "label" = @Translation("Apigee Edge endpoint"),
+ *         "required" = false
+ *       },
+ *       "authorization_server" = {
+ *         "label" = @Translation("Authorization server"),
+ *         "required" = false
+ *       },
+ *       "client_id" = {
+ *         "label" = @Translation("Client ID"),
+ *         "required" = false
+ *       },
+ *       "client_secret" = {
+ *         "label" = @Translation("Client secret"),
+ *         "required" = false
+ *       }
+ *     }
+ *   }
+ * )
+ */
+class ApigeeAuthKeyType extends EdgeKeyTypeBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function generateKeyValue(array $configuration) {
+    return '[]';
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function validateKeyValue(array $form, FormStateInterface $form_state, $key_value) {
+    if (empty($key_value)) {
+      return;
+    }
+
+    $value = $this->unserialize($key_value);
+    if ($value === NULL) {
+      $form_state->setError($form, $this->t('The key value does not contain valid JSON: @error', ['@error' => json_last_error_msg()]));
+      return;
+    }
+
+    foreach ($this->getPluginDefinition()['multivalue']['fields'] as $id => $field) {
+      if (isset($field['required']) && $field['required'] === FALSE) {
+        continue;
+      }
+
+      $error_element = $form['settings']['input_section']['key_input_settings'][$id] ?? $form;
+
+      /** @var \Drupal\Core\StringTranslation\TranslatableMarkup $field */
+      if (!isset($value[$id])) {
+        $form_state->setError($error_element, $this->t('The key value is missing the field %field.', ['%field' => $field['label']->render()]));
+      }
+      elseif (empty($value[$id])) {
+        $form_state->setError($error_element, $this->t('The key value field %field is empty.', ['%field' => $field['label']->render()]));
+      }
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getAuthenticationMethod(KeyInterface $key): Authentication {
+    $values = $key->getKeyValues();
+    if ($values['auth_type'] === EdgeKeyTypeInterface::EDGE_AUTH_TYPE_OAUTH) {
+      // Use Oauth authentication.
+      return new OauthAuthentication($this->getUsername($key), $this->getPassword($key), \Drupal::service('apigee_edge.token_storage'), NULL, $this->getClientId($key), $this->getClientSecret($key), NULL, $this->getAuthorizationServer($key));
```

Comment:
Inject it with the overriding `create()` method instead of calling `\Drupal::service()` directly.

Reply:
The service is not used unless `auth_type` == `oauth`.